File: ios/Classes/Map/PolylinesController.swift
//
// Created by Valentin Grigorean on 28.04.2021.
//

import Foundation
import ArcGIS

class PolylinesController: NSObject, SymbolsController {

    private var polylineIdToController = Dictionary<String, PolylineController>()

    private let graphicsOverlays: GraphicsOverlay

    private let methodChannel: FlutterMethodChannel

    var selectionPropertiesHandler: SelectionPropertiesHandler?

    var symbolVisibilityFilterController: SymbolVisibilityFilterController?

    init(methodChannel: FlutterMethodChannel,
         graphicsOverlays: GraphicsOverlay) {
        self.methodChannel = methodChannel
        self.graphicsOverlays = graphicsOverlays
    }

    func addPolylines(polylinesToAdd: [Dictionary<String, Any>]) {
        for polyline in polylinesToAdd {
            let polylineId = polyline["polylineId"] as! String
            let controller = PolylineController(polylineId: polylineId)
            polylineIdToController[polylineId] = controller
            updatePolyline(data: polyline, controller: controller)
            controller.add(graphicsOverlay: graphicsOverlays)
        }
    }

    func changePolylines(polylinesToChange: [Dictionary<String, Any>]) {
        for polyline in polylinesToChange {
            let polylineId = polyline["polylineId"] as! String
            guard let controller = polylineIdToController[polylineId] else {
                continue
            }
            updatePolyline(data: polyline, controller: controller)
        }
    }

    func removePolylines(polylineIdsToRemove: [String]) {
        for polylineId in polylineIdsToRemove {
            guard let controller = polylineIdToController[polylineId] else {
                continue
            }
            controller.remove(graphicsOverlay: graphicsOverlays)
            polylineIdToController.removeValue(forKey: polylineId)
        }
    }

    private func updatePolyline(data: Dictionary<String, Any>,
                                controller: PolylineController) {
        updateController(controller: controller, data: data)

        if let color = UIColor(data: data["color"]) {
            controller.setColor(color: color)
        }

        if let styleIndex = data["style"] as? Int {
            controller.setStyle(style:  SimpleLineSymbol.Style(styleIndex))
        }

        if let pointsData = data["points"] as? [Dictionary<String, Any>] {
            var points: [Point] = []
            for data in pointsData {
                points.append(Point(data: data))
            }
            controller.setPoints(points: points)
        }

        if let width = data["width"] as? Double {
            controller.setWidth(width: CGFloat(width))
        }

        if let antialias = data["antialias"] as? Bool {
            controller.setAntialias(antialias: antialias)
        }
    }
}

extension PolylinesController: MapGraphicTouchDelegate {

    func canConsumeTaps() -> Bool {
        for (_, controller) in polylineIdToController {
            if controller.consumeTapEvents {
                return true
            }
        }
        return false
    }

    func didHandleGraphic(graphic: Graphic) -> Bool {
        guard let polylineId = graphic.attributes["polylineId"] as? String else {
            return false
        }
        if let currentPolyline = polylineIdToController[polylineId] {
            guard currentPolyline.consumeTapEvents else {
                return false
            }
            methodChannel.invokeMethod("polyline#onTap", arguments: ["polylineId": polylineId])
        }
        return true
    }
}

File: android/src/main/kotlin/com/valentingrigorean/arcgis_maps_flutter/mapping/symbology/PolylinesController.kt
package com.valentingrigorean.arcgis_maps_flutter.mapping.symbology

import com.arcgismaps.mapping.view.Graphic
import com.arcgismaps.mapping.view.GraphicsOverlay
import com.valentingrigorean.arcgis_maps_flutter.convert.map.toPolylineIdValue
import com.valentingrigorean.arcgis_maps_flutter.map.MapTouchGraphicDelegate
import io.flutter.plugin.common.MethodChannel

class PolylinesController(
    private val methodChannel: MethodChannel,
    private val graphicsOverlay: GraphicsOverlay
) : BaseSymbolController(), MapTouchGraphicDelegate {
    private val polylineIdToController: MutableMap<String?, PolylineController> = HashMap()
    override fun canConsumeTaps(): Boolean {
        for (controller in polylineIdToController.values) {
            if (controller.consumeTapEvents) {
                return true
            }
        }
        return false
    }

    override fun didHandleGraphic(graphic: Graphic): Boolean {
        val rawPolylineId = graphic.attributes["polylineId"] ?: return false
        val polylineId = rawPolylineId as String
        val controller = polylineIdToController[polylineId]
        if (controller == null || !controller.consumeTapEvents) {
            return false
        }
        methodChannel.invokeMethod("polyline#onTap", polylineId.toPolylineIdValue())
        return true
    }

    fun addPolylines(polylinesToAdd: List<Any>?) {
        if (polylinesToAdd == null) {
            return
        }
        val graphicsAdded = ArrayList<Graphic>(polylinesToAdd.size)
        for (polyline in polylinesToAdd) {
            val data = polyline as Map<*, *>? ?: continue
            val polylineId = data["polylineId"] as String
            val controller = PolylineController(polylineId)
            controller.setSelectionPropertiesHandler(selectionPropertiesHandler)
            controller.interpretGraphicController(data, symbolVisibilityFilterController)
            polylineIdToController[polylineId] = controller
            graphicsAdded.add(controller.graphic)
        }
        graphicsOverlay.graphics.addAll(graphicsAdded)
    }

    fun changePolylines(polylinesToChange: List<Any>?) {
        if (polylinesToChange == null) {
            return
        }
        for (polyline in polylinesToChange) {
            val data = polyline as Map<*, *>? ?: continue
            val polylineId = data["polylineId"] as String?
            polylineIdToController[polylineId]?.interpretGraphicController(
                data,
                symbolVisibilityFilterController
            )
        }
    }

    fun removePolylines(polylineIdsToRemove: List<Any>?) {
        if (polylineIdsToRemove == null) {
            return
        }
        val graphicsToRemove = ArrayList<Graphic>(polylineIdsToRemove.size)
        for (rawPolylineId in polylineIdsToRemove) {
            val polylineId = rawPolylineId as String
            val controller = polylineIdToController.remove(polylineId)
            if (controller != null) {
                onSymbolRemoval(controller)
                graphicsToRemove.add(controller.graphic)
            }
        }
        graphicsOverlay.graphics.removeAll(graphicsToRemove)
    }
}